3.next - Add client compatibilty methods to CookieCollection

markstory opened the pull request:
Add the methods that make Http\Cooke\CookieCollection compatible with Http\Client. Next, I'll use the new class in the client.

I've also fixed a few issues around how cookies were being stored. Cookies need to be unique per name;domain;path tuple. Without this constraint expiring cookies becomes very challenging. Also it becomes impossible to update a cookie in the collection.

Refs #10406

markstory committed:
Add Cookie:getId()

Having identifiers will allow collections to replace cookie values
when updating collections from responses.

markstory committed:
Add store() and fix cookie expiration/overwrite

Add store() method for backwards compat with Client\CookieCollection.
I've also fixed a few issues that resulted in:

1. Cookies not be updated when a new response has a new value.
2. Cookies not being removed when they expire.

Both of these issues have been solved with an additional private method.

markstory committed:
Add backwards compatibility mode for get()

This is somewhat ugly code, but it'll be going away in 4.0.0.

markstory committed:
Remove code that is implemented in the parent class.

These methods now exist in the base class.

saeideng commented on src/Http/Cookie/CookieCollection.php:
```diff
@@ -315,4 +361,81 @@ protected function parseSetCookieHeader($values)
+     * @param \Cake\Http\Client\Response $response The response to read cookies from
+     * @param string $url The request URL used for default host/path values.
+     * @return void
+     * @deprecated 3.5.0 Will be removed in 4.0.0. Use `addFromResponse()` instead.
```
CookieCollection.php **created** and **deprecated** in 3.5.0 before release?

saeideng replied:
CookieCollection.php **created** and **deprecated** in 3.5.0 ,before release?

saeideng commented on src/Http/Cookie/CookieCollection.php:
```diff
@@ -315,4 +361,81 @@ protected function parseSetCookieHeader($values)
+     * This method should not be used and is only provided for backwards compatibility.
+     *
+     * @return array
+     * @deprecated 3.5.0 Will be removed in 4.0.0
```
CookieCollection.php **created** and **deprecated** in 3.5.0 ,before release?

markstory replied:
That is correct for this method. I want to replace the `Http\Client\CookieCollection` with this new class. In order to do that, this new class needs to provide the same public methods as the old class. However, I feel that the old methods are a bit clunky in both name and parameters, so I would like to deprecate them.

Another option I considered was to modify `Http\Client\CookieCollection` to extend this new class and leave the backwards compatibility there. I'm still open to that approach if people think it would be easier to understand in the future.

saeideng commented on src/Http/Cookie/Cookie.php:
```diff
@@ -212,6 +212,18 @@ public function withName($name)
+     */
+    public function getId()
+    {
+        return "{$this->name};{$this->domain};{$this->path}";
```
needs here  a `mb_strtolower()` ?

saeideng commented on src/Http/Cookie/CookieCollection.php:
```diff
@@ -103,6 +106,22 @@ public function get($name)
+     *
+     * @param string $url The url to get cookies for.
+     * @return array An array of matching cookies.
+     * @deprecated 3.5.0 Will be removed in 4.0.0. Use addToRequest() instead.
```
**deprecated** even for protected method?
CookieCollection.php **created** and **deprecated** in 3.5.0 ,before release?

markstory replied:
Yes, see my comment below.

saeideng replied:
anyway this method (` protected function getByUrl()`) is with protected scope and no need to this line

markstory replied:
Ok.

saeideng commented on src/Http/Cookie/RequestCookies.php:
```diff
@@ -37,33 +37,4 @@ public static function createFromRequest(ServerRequestInterface $request, $cooki
-     * @param string $name Name of the cookie
-     * @return bool
-     */
-    public function has($name)
```
probably needs **deprecated** instead of removing !?

markstory replied:
No, because this method was never part of any actual release.

saeideng commented on src/Http/Cookie/RequestCookies.php:
```diff
@@ -37,33 +37,4 @@ public static function createFromRequest(ServerRequestInterface $request, $cooki
-     * @throws \InvalidArgumentException
-     * @return Cookie
-     */
-    public function get($name)
```
probably needs **deprecated** instead of removing !?

markstory replied:
No, because this method was never part of any actual release.

markstory replied:
Also both this method and `has()` are implemented in this class's parent. It is worth recognizing that this class will be removed very soon as well.

markstory committed:
cookie ids should be lower case.

Also remove deprecation warning as protected methods don't really need
these markers.

markstory committed:
Fix phpcs error

antograssiot commented on src/Http/Cookie/CookieCollection.php:
```diff
@@ -65,24 +64,25 @@ public function count()
      * Get the first cookie by name.
      *
-     * If the provided name matches a URL (matches `#^https?://#`) this method
+     * If the provided name matches a URL (starts with `http:`) this method
```
the `:`should be there I think

markstory commented:
I've been thinking about this some more. I'm going to push all the backwards compatibility methods into the `Http\Client\CookieCollection` and extending the new class. I think this will keep the new code cleaner/simpler.

markstory committed:
Move deprecated code into the deprecated client CookieCollection.

This allows the 'new' collection to stay simpler and cleaner.

saeideng commented on src/Http/Cookie/CookieCollection.php:
```diff
@@ -14,6 +14,7 @@
 namespace Cake\Http\Cookie;
 
 use ArrayIterator;
+use Cake\Http\Client\Response as ClientResponse;
```
you can remove this line :)

saeideng commented:
extending class and use of 
` Cake\Http\CookieCollection::$cookies` in `Cake\Http\Client\CookieCollection`
means ,we have some cookie in both Collection ! ? :worried:

markstory commented:
@saeideng I don't understand your question. There should only be one list of cookies in both collections?
